torch_geometric/nn/aggr/base.py

```diff
@@ -7,13 +7,53 @@
 from torch_geometric.utils import to_dense_batch
 
 
+def validate(x: Tensor, index: Optional[Tensor] = None,
+             ptr: Optional[Tensor] = None, dim_size: Optional[int] = None,
+             dim: int = -2):
+
+    if dim >= x.dim() or dim < -x.dim():
+        raise ValueError(f"Encountered invalid dimension '{dim}' of "
+                         f"source tensor with {x.dim()} dimensions")
+
+    if index is None and ptr is None:
+        index = x.new_zeros(x.size(dim), dtype=torch.long)
+
+    if ptr is not None:
+        if dim_size is None:
+            dim_size = ptr.numel() - 1
+        elif dim_size != ptr.numel() - 1:
+            raise ValueError(f"Encountered invalid 'dim_size' (got "
+                             f"'{dim_size}' but expected "
+                             f"'{ptr.numel() - 1}')")
+
+    if index is not None:
+        if dim_size is None:
+            dim_size = int(index.max()) + 1 if index.numel() > 0 else 0
+        elif index.numel() > 0 and dim_size <= int(index.max()):
+            raise ValueError(f"Encountered invalid 'dim_size' (got "
+                             f"'{dim_size}' but expected "
+                             f">= '{int(index.max()) + 1}')")
+
+    return x, index, ptr, dim_size, dim
+
+
 class Aggregation(torch.nn.Module):
     r"""An abstract base class for implementing custom aggregations."""
+    def __init__(self):
+        super().__init__()
+        self._forward_sub = self.forward
+        self.forward = self._foward
```

Comment:
I think this is overly complicated. How about we just do
```
def forward:
    self.validate()
    self._forward()
```
And overwrite `_forward` in child modules?

Reply:
yeah I was thinking that too - just providing this option incase we want to keep the  current abstract interface